README
NAME
    Captive::Portal - Perl based solution for controlled network access

ABSTRACT
    A so called *Captive Portal* written in perl for Linux Gateways. For a
    longer explanation see:

    <http://en.wikipedia.org/wiki/Captive_Portal>

DESCRIPTION
    Captive::Portal a.k.a. CaPo is a hotspot solution for Linux Gateways.
    CaPo is developed and in service at Ulm University for thousands of
    concurrent users.

    The main focus is scalability, performance, simple administration and
    user-friendliness. The goals have been achieved by using scalable
    technologies like ipset/iptables, FastCGI/CGI and a fine tuned
    concurrent session handling based on the filesystem locking mechanism
    without any need for an additional RDBMS.

    CaPo is compatible with any FastCGI/CGI enabled HTTPS-server.

ALGORITHM IN SHORT
    1. Access Denied
        Only selected protocols like DHCP/DNS/NTP/IMAPS/IPSec/... to
        selected servers are allowed for unauthenticated clients.

    2. Internal NAT redirect
        HTTP-traffic on the gateways inside interface - from unauthenticated
        clients - is redirected by an iptables(8) NAT-rule to a port the
        HTTP-server is listening, e.g.

         iptables -t nat -A PREROUTING -i eth1 -p tcp --dport 80 -j REDIRECT --to-port 5281

    3. External HTTP redirect
        The HTTP-server redirects the HTTP-request by a rewrite rule to an
        HTTPS-request for the CaPo script *capo.fcgi* , e.g.

         <VirtualHost *:5281>
             RewriteEngine On
             RewriteRule   .*  https://gateway.acme.org/capo/? [R,L]
         </VirtualHost>

         CLIENT REQUEST:

            GET /foo/bar HTTP/1.1
            Host: www.test.org

         SERVER RESPONSE:

            HTTP/1.1 302 Found
            Location: https://gateway.acme.org/capo/

    4. SESSION LOGIN
        The *capo.fcgi* script, fired due to this redirected request, offers
        a splash/login page. After successful login the firewall is
        dynamically changed to allow this clients IP/MAC tuple for internet
        access by ipset(8):

            ipset --add capo_sessions_ipset CLIENT_IP,CLIENT_MAC

    5. SESSION LOGOUT
        The capo.fcgi script offers a status/logout page. After successful
        logout the firewall is dynamically changed to disallow this IP/MAC
        tuple for internet access.

            ipset --del capo_sessions_ipset CLIENT_IP

    6. SESSION IDLE
        A cronjob fires periodically the capo-ctl.pl script checking for
        idle sessions. Idle means, the client didn't send any packet for a
        period of time (cfg param: IDLE_TIME = 10min). Before a session is
        put into IDLE state the client is once pinged.

        It is a design goal not requiring JavaScript on clients!

    7. COMFORTABLE SESSION REACTIVATION
        For a short period of time (cfg param: KEEP_OLD_STATE_PERIOD = 1h)
        the session is still on disc, but in IDLE state. If a client has
        cookies enabled and a HTTP request matches the stored IP/MAC/COOKIE
        data on disc, the session is reactivated without a login page.

INSTALLATION
    Please see the INSTALL file in this distribution. As a minimum please be
    aware of the following access restrictions:

    Captive::Portal needs access to iptables(8) and ipset(8) to change the
    firewall-rules on request. You must add the following rule (or similar)
    to the sudoers file, depending on the username of your http daemon:

     WWW_USER ALL=NOPASSWD: /PATH/TO/iptables, /PATH/TO/ipset

    If you use fping(8) (see USE_FPING config parameter) to trigger idle
    sessions before going idle you must add fping to the sudoers file like
    ipset and iptables, regardless of the suid bit on fping, since we need
    special timing flags available only for root:

     WWW_USER ALL=NOPASSWD: /PATH/TO/iptables, /PATH/TO/ipset, /PATH/TO/fping

    The default $SESSIONS_DIR is set to '/var/cache/capo'.

    WWW_USER must be the owner of this dir with write permissions!

CONFIGURATION
    The configuration file is searched in the following default places:

        $ENV{CAPTIVE_PORTAL_CONFIG} ||
        $Bin/../etc/local/config.pl ||
        $Bin/../etc/config.pl

LOGGING
    Logging is handled by the Log::Log4perl module. The logging
    configuration is searched in the following default places:

        $ENV{CAPTIVE_PORTAL_LOG4PERL}   ||
        $Bin/../etc/local/log4perl.conf ||
        $Bin/../etc/log4perl.conf

LOCAL ADAPTATION
    The HTML files are generated from templates (Template-Toolkit syntax).
    You should use the original template files as stanzas and put the
    locally changed versions into the local template tree. The template
    search order prefers the local templates.

    The CSS is based on the wonderful blueprint css framework, see
    <http://www.blueprintcss.org/>. Of course you may use your own styles if
    needed.

    The firewall rules and commands are also generated from template files.
    Normally there is no need to change the firewall rules but it would be
    possible to add some local needed additional rules without changing the
    perl code. Be careful, you must understand the algorithm and the
    different states. Best you ask the author for any modifications.

I18N
    There exists a template tree for each supported language. Some
    system/error messages used by the program must also be translated for
    the message catalog in the config file.

SEE ALSO
    Captive::Portal::Role::Config, capo.fcgi and capo-ctl.pl

CREDITS
    Most of the good parts have been implemented by many creative discussion
    with my colleague Bernd Leibing.

BUGS AND LIMITATIONS
    There are no known problems with this module.

    Please report any bugs or feature requests to "bug-captive-portal at
    rt.cpan.org", or through the web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Captive-Portal>. I will
    be notified, and then you'll automatically be notified of progress on
    your bug as I make changes.

AUTHOR
    Karl Gaissmaier, "<gaissmai at cpan.org>"

LICENSE AND COPYRIGHT
    Copyright 2010-2012 Karl Gaissmaier, all rights reserved.

    This distribution is free software; you can redistribute it and/or
    modify it under the terms of either:

    a) the GNU General Public License as published by the Free Software
    Foundation; either version 2, or (at your option) any later version, or

    b) the Artistic License version 2.0.



Changes
Revision history for Captive::Portal

4.09    Wed, 13 Mar 2013 11:29:04 +0100
	* get rid of SUID, use capabilities

4.07    Mon, 11 Mar 2013 16:54:10 +0100
	* get rid of sudo, use SUID for iptables and ipset

4.06    Tue, 26 Feb 2013 17:51:34 +0100
	* fw_start_session also sets IP in the capo_activity_ipset
          needed for stateful restarts

4.05    Tue, 26 Feb 2013 16:42:41 +0100
	* changed local templates for status view
	* reversed template vars to prior 4.03

4.04    Tue, 26 Feb 2013 16:08:00 +0100
	* changed templates for status view
	* fixed syntax bug from 4.03 in capo-ctl.pl

4.03    Mon, 25 Feb 2013 15:46:44 +0100
	* minor improvements

4.02    Mon, 25 Feb 2013 11:57:46 +0100
	* status view: show number of clients behind firewall from capo_activity_ipset
	* minor pod corrections

4.01    Thu, 21 Feb 2013 20:40:10 +0100
	* bump COPYRIGHT to ..2013
	* minor bug in parsing 'ipset list ...' line for 0 values in timeout

4.00    Thu, 21 Feb 2013 17:54:19 +0100
	* incompatible changes, no more support for ipset < 4.x
	* use timeout in ipsets for IDLE algorithm
	* allow incoming traffic only for active clients, probibit ARP sweeps
	* capo_activity_ipset is now bitmap:ip and no longer bitmap:ip,mac due to timeout bug in current ipset
          

3.22    Wed, 20 Feb 2013 16:51:49 +0100
	* repair bug for incoming traffic for not yet authenticated clients
	* repair bug if --add-set in iptables not using the default timeout of the ipset

3.21    Thu, 29 Nov 2012 14:50:52 +0100
	* no more support for ipset() < 4.x
	* use timeout option in ipset() to detect idle clients
	* no more support for fping

3.16-20 Thu, 29 Nov 2012 10:25:08 +0100
	* only Uni-Ulm local changes

3.15    Wed, 25 Apr 2012 11:51:50 +0200
	* use JS to logg sslstrip attacks

3.12    Thu, 02 Feb 2012 12:12:56 +0100
	* use Spawn::Safe instead of Capture::Tiny

3.11    Mon, 30 Jan 2012 09:23:34 +0100
	* fixed typo in cfg slot for ipset_version

3.10    Sun, 29 Jan 2012 15:39:34 +0100
	* merged version 2 and version 3 again together
	* reap zombies in capo.fcgi after each HTTP request

3.03    Thu, 26 Jan 2012 10:50:13 +0100
	* more DEBUG in authentication
	* LOG in filter.tt firewall template removed
	* mangled searchpath for ipset

3.02    Wed, 25 Jan 2012 17:50:14 +0100
	* call fw_status less often

3.01    Mon, 23 Jan 2012 11:15:42 +0100
	* rewrite for newer ipset versions greater 5.x
	* add hint for version mismatch in INSTALL file
	* fix pod2usage due to $0 hack
	* use Captive::Portal Version 2.x line for ipset < version 5.x

2.28    Fri, 20 Jan 2012 11:02:07 +0100
	* fixed pod2usage error in capo-ctl.pl

2.27    Thu, 19 Jan 2012 21:33:25 +0100
	* hint for ipset/capo version mismatch in INSTALL file

2.26    Thu, 19 Jan 2012 11:06:13 +0100
	* arrgh, ipset API changed, check for ipset major version > 4 in Build.PL
	* fixed minor POD typos

2.25    Mon, 07 Nov 2011 10:13:15 +0100
	* fixed again POD typos
	* added script_files in Build.pl
	* put scriptname for capo-ctl.pl in process table

2.24    Fri, 04 Nov 2011 10:23:05 +0100
	* fixed POD error in capo-ctl.pl

2.23    Wed, 02 Nov 2011 11:03:51 +0100
	* again more locking tweaks in handling old, idle sessions

2.22    Mon, 31 Oct 2011 11:09:42 +0100
	* still more timing tweaks in deleting old, idle sessions

2.21    Sat, 29 Oct 2011 19:57:44 +0200
	* more small tweaks in deleting old, idle sessions

2.20    Fri, 28 Oct 2011 15:34:43 +0200
	* solved race condition in deleting old, idle sessions

2.19	Tue, 25 Oct 2011 09:48:12 +0200
	* removed unneeded modules for locking from Captive::Portal::Role::Utils

2.18	Tue, 25 Oct 2011 09:36:23 +0200
	* OS hint in Build.pl for CPAN testers
        * fixed POD error in Captive::Portal::LockHandle

2.17	Wed, 12 Oct 2011 15:14:56 +0200
	* use Captive::Portal::LockHandle as Class, no longer MIXIN get_lock_handle()

2.16	Fri, 07 Oct 2011 09:52:45 +0200
	* changed locking from being a role to a class for DESTROY() handling
	* documentation improvements
	* test improvements

2.15	Thu, 06 Oct 2011 12:12:00 +0200
	* capo.fcgi: put scriptname in process table

2.14	Thu Oct 6 10:52:42 2011 +0200
	* fixed minor documentation bugs

2.13	Wed Oct 5 17:55:07 2011 +0200
	* minor bug, Scalar::Util in role I18N missing

2.12	Tue Sep 27 10:15:16 2011 +0200
	* more documentation fixes

2.11	Mon Sep 26 14:36:49 2011 +0200
	* fixed minor documentation bugs

2.10	Fri Sep 23 16:23:17 2011 +0200
	* initial CPAN release, already in production for months



lib/Captive/Portal/Role/Config.pm
package Captive::Portal::Role::Config;

use strict;
use warnings;

=head1 NAME

Captive::Portal::Role::Config - config reader for Captive::Portal

=head1 DESCRIPTION

Config file parser and storage for cfg hash. The configuration syntax is perl.


=cut

our $VERSION = '4.09';

use Log::Log4perl qw(:easy);
use FindBin qw($Bin);
use File::Spec::Functions qw(splitdir rootdir catfile catdir);

use Role::Basic;

# just bin/../ => bin
my @bin_parts = splitdir($Bin);
pop(@bin_parts);

use constant TRUE  => 1;
use constant FALSE => 0;

use constant ON  => 1;
use constant OFF => 0;

use constant YES => 'yes';
use constant NO  => '';

use vars qw($APP_NAME $APP_DIR);

$APP_NAME = 'capo';
$APP_DIR  = catdir(@bin_parts);

# SINGLETON
my $cfg_hash = {};

=head1 PRESET GLOBAL PACKAGE VARIABLES

The following variables are predefined and can be used for interpolation in config values.

 $APP_NAME = 'capo'

 $APP_DIR = "$Bin/../"

=head1 PRESET DEFAULTS

=over 4

=item DOCUMENT_ROOT => "$APP_DIR/static"

Basedir for static content like images, css or error pages.

=item TEMPLATE_INCLUDE_PATH => "$APP_DIR/templates/local/:$APP_DIR/templates/orig"

Directories to search for templates.

=item RUN_USER => 'wwwrun'

Drop privileges to RUN_USER.

=item RUN_GROUP => 'www',

Drop privileges to RUN_GROUP.

=item SESSIONS_DIR => "/var/cache/$APP_NAME"

Where to store the session files. This directory must exist und must be readable/writeable by RUN_USER.

=item SSL_REQUIRED => ON

A JS script looks for SSL encryption of the login/splash page and throws an error when not. Maybe a man-in-the-middle plays http-https proxy like sslstrip(8). If the mitm strips JS then this doesn't help anyway. The users must check the location bar for HTTPS these days, sigh.

=item SESSION_MAX => 48 * 3600    # 2d

Max session time until a forced disconnect.

=item IDLE_TIME => 60 * 10      # 10 min

How long to wait for activity from ip/mac until a session is marked idle.

=item KEEP_OLD_STATE_PERIOD => 1 * 60 * 60,  # 1h

How long to keep idle session records on disk for fast reconnect with proper ip/mac/cookie match.

=back

=head1 LOCAL PARAMETERS

=over 4

=item ADMIN_SECRET

Passphrase for detailed sessions view.

=item AUTHEN_SIMPLE_MODULES

Authentication is handled by the Authen::Simple framework. You may stack any of the Authen::Simple::... plugins for authentication, see the $Bin/../etc/config.pl template.

=item IPTABLES->capture_if => 'eth1'

The inside gateway interface, e.g. 'eth1'. All http traffic, not allowed by any predefined rule, is captured and redirected to the capo.fcgi script.

=item IPTABLES->capture_net => '192.168.0.0/22'

The inside IP network in CIDR notation, e.g. '192.168.0.0/22'

=item IPTABLES->capture_ports => [80, 8080]

What tcp ports should be captured and redirected, e.g. [ 80, 8080]

=item IPTABLES->redirect_port => 5281

The port where the HTTP-server is listen in order to rewrite this http request to an https request.

The above settings result in a NAT rule equivalent to:

 iptables -t nat -A PREROUTING -i eth1 -s 192.168.0.0/22 ! -d 192.168.0.0/22 \
          -p tcp -m multiport --dports 80,8080 -j  REDIRECT --to-port 5281

=item IPTABLES->throttle => OFF

You may throttle HTTP/HTTPS requests/sec per client IP. Some clients/gadgets fire a lot of HTTP traffic without human intervention. Depending on your hardware and your encryption resources this will overload your gateway.

=item IPTABLES->throttle_ports => [ 80, 5281]

You should protect/throttle port 80 and the redirect_port (see above).


=item IPTABLES->throttle_seconds => 30

=item IPTABLES->throttle_hitcount => 15

Both parameters define the average and the burst. Average is hitcount/seconds and burst is hitcount in seconds. With the values of 30 and 15, the average would be 15hits/30s => 1hit/2s. The burst would be 15hits in 30 seconds.

The above settings result in iptable rules equivalent to:

 # throttle/drop new connections
 iptables -t filter -A INPUT -p tcp --syn -m multiport --dports 80,5281 \
    -m recent --name capo_throttle --rcheck --seconds 30 --hitcount 15 -j DROP

 # at last accept new connections but set/update the recent table
 iptables -t filter -A INPUT -p tcp --syn -m multiport --dports 80,5281 \
    -m recent --name capo_throttle --set -j ACCEPT

=item IPTABLES->open_services

Allow access to open local services like DHCP, DNS, NTP, ...

=item IPTABLES->open_clients

Allow access for some dumb clients without authentication.

=item IPTABLES->open_servers

Allow access to some open servers.

=item IPTABLES->open_networks

Allow access to some open networks.

=item I18N_LANGUAGES

Supported languages for system messages and HTML templates.

=item I18N_FALLBACK_LANG

Fallback language if the client message isn't supported in the system message catalog and templates.

=item I18N_MSG_CATALOG

Translations of the system messages.

=back

=cut

my %pre_defaults = (
    DOCUMENT_ROOT => catdir( $APP_DIR, 'static' ),

    TEMPLATE_INCLUDE_PATH => catdir( $APP_DIR, 'templates', 'local' ) . ':'
      . catdir( $APP_DIR, 'templates', 'orig' ),

    SESSIONS_DIR => catdir( rootdir(), 'var', 'cache', $APP_NAME ),

    RUN_USER  => 'wwwrun',
    RUN_GROUP => 'www',

    SSL_REQUIRED          => ON,
    SESSION_MAX           => 2 * 24 * 60 * 60,    # 2 days
    KEEP_OLD_STATE_PERIOD => 1 * 60 * 60,         # 1h

    IDLE_TIME     => 10 * 60,                     # 10min before set to idle

    I18N_LANGUAGES     => [ 'en', ],
    I18N_FALLBACK_LANG => 'en',
);

# Role::Basic exports ALL subroutines, there is currently no other way to
# prevent exporting private methods, sigh
#
my ($_priv_post_defaults, $_priv_check_cfg);

=head1 ROLES

=over

=item $capo->parse_cfg_file($filename)

Parse config file, merge with defaults. Die on error.

=cut

sub parse_cfg_file {
    my $self     = shift;
    my $cfg_file = shift;
    LOGDIE "missing parameter 'config_file'" unless defined $cfg_file;

    DEBUG "preset cfg_hash with default values";
    $cfg_hash = {%pre_defaults};

    DEBUG "parse config file $cfg_file";
    my $parsed_cfg_file = do $cfg_file;

    # check the config file for syntactic errors
    LOGDIE "couldn't parse $cfg_file: $@" if $@;
    LOGDIE "couldn't do $cfg_file: $!"
      unless defined $parsed_cfg_file;
    LOGDIE "couldn't run $cfg_file" unless $parsed_cfg_file;

    DEBUG "merge parsed values with preset default values to cfg_hash";
    $cfg_hash = { %$cfg_hash, %$parsed_cfg_file };

    $self->$_priv_check_cfg();

    $self->$_priv_post_defaults();

    return 1;
}

=item $capo->cfg()

Getter, return a shallow copy of the config hashref.

=cut

sub cfg { return {%$cfg_hash}; }

#
# Add some defaults after reading cfg file. Must be postponed to
# interpolate of already set params.
#

$_priv_post_defaults = sub {
    my $self     = shift;

    # defined as anonymous sub,
    # else Role::Basic would export this as role, sigh!

    DEBUG "add post_parse config default values, if needed";

    unless ( exists $cfg_hash->{LOCK_FILE} ) {
        $cfg_hash->{LOCK_FILE} =
          catfile( $cfg_hash->{SESSIONS_DIR}, 'capo-ctl.lock' );
    }
};

#
# semantic params validation of cfg_hash
#

$_priv_check_cfg = sub {

    # defined as anonymous sub,
    # else Role::Basic would export this as role, sigh!

    DEBUG "do cfg_hash params validation";

    # check the config file for sematic errors and warnings
    if ( $cfg_hash->{BOILERPLATE} ) {
        LOGDIE 'FATAL: the config file is in BOILERPLATE state';
    }

    unless ( $cfg_hash->{SESSIONS_DIR} ) {
        LOGDIE 'FATAL: missing SESSIONS_DIR in cfg file';
    }

    if ( $cfg_hash->{MOCK_MAC} ) {
        ERROR "uncomment 'MOCK_MAC' for production in cfg file";
    }

    if ( $cfg_hash->{MOCK_FIREWALL} ) {
        ERROR "uncomment 'MOCK_FIREWALL' for production in cfg file";
    }

    if ( $cfg_hash->{MOCK_AUTHEN} ) {
        ERROR "uncomment 'MOCK_AUTHEN' for production in cfg file";
    }
    else {
        ERROR 'missing Authen::Simple modules in cfg file'
          unless $cfg_hash->{'AUTHEN_SIMPLE_MODULES'};
    }

    unless ( $cfg_hash->{ADMIN_SECRET} ) {
        ERROR 'missing ADMIN_SECRET in cfg file';
    }

    unless ( $cfg_hash->{SSL_REQUIRED} ) {
        ERROR 'set SSL_REQUIRED for production in cfg file';
    }

    unless ( $cfg_hash->{IPTABLES}{capture_if} ) {
        ERROR "missing 'capture_if' in cfg file";
    }

    unless ( $cfg_hash->{IPTABLES}{capture_net} ) {
        ERROR "missing 'capture_net' in cfg file";
    }

    unless ( $cfg_hash->{IPTABLES}{capture_ports} ) {
        ERROR "missing 'capture_ports' in cfg file";
    }

    unless ( $cfg_hash->{IPTABLES}{redirect_port} ) {
        ERROR "missing 'redirect_port' in cfg file";
    }
};

1;

=back

=head1 AUTHOR

Karl Gaissmaier, C<< <gaissmai at cpan.org> >>

=head1 LICENSE AND COPYRIGHT

Copyright 2010-2013 Karl Gaissmaier, all rights reserved.

This distribution is free software; you can redistribute it and/or modify it
under the terms of either:

a) the GNU General Public License as published by the Free Software
Foundation; either version 2, or (at your option) any later version, or

b) the Artistic License version 2.0.

=cut

#vim: sw=4
